ppc64le: eigen: using the new flag to enable MMA dynamic dispatch

maxiwell opened the pull request:
Eigen removed the dynamic dispatch by default. Now, it's necessary to add
EIGEN_ALTIVEC_ENABLE_MMA_DYNAMIC_DISPATCH=1 to enable this feature.

https://gitlab.com/libeigen/eigen/-/commit/7b10795e3939518f1be24f5799e01dbaa1e0c04e
https://gitlab.com/libeigen/eigen/-/commit/591906477bc8c8102dbefceefe10d81648865394

cantonios commented:
Hi @maxiwell, I just want to confirm this is actually what IBM wants.  It was removed by default because it's broken on system-installed compilers on many systems (e.g. ubuntu), since it requires a specific ld version that we cannot detect from within Eigen.  The thinking was that build systems might be able to detect this and turn on the flag if desired.

maxiwell commented:
Hi @cantonios, thanks for your reply.

> I just want to confirm this is actually what IBM wants. 


In this case, do you know a good way to check the ld version here? Maybe changing the configure.py?

maxiwell commented:
Hi @cantonios, Thanks for your reply

> I just want to confirm this is actually what IBM wants.

Can I help with this task?

> The thinking was that build systems might be able to detect this and turn on the flag if desired.

In this case, maybe we can change the `configure.py` to check the ld version and add the flag. Is it a good way?

cantonios commented:
> In this case, maybe we can change the `configure.py` to check the ld version and add the flag. Is it a good way?

I have no idea, this is not my area of expertise.  But I do think something like this would be preferable over just always blindly adding the flag.  Otherwise, people who try to build on ppc+ubuntu will get compile errors.

maxiwell committed:
ppc64le: eigen: using the new flag to enable MMA dynamic dispatch if ld >= 2.35

Eigen removed the dynamic dispatch by default. Now, it's necessary to add
EIGEN_ALTIVEC_ENABLE_MMA_DYNAMIC_DISPATCH=1 to enable this feature if
the linker version is 2.35 or greater.

https://gitlab.com/libeigen/eigen/-/commit/7b10795e3939518f1be24f5799e01dbaa1e0c04e
https://gitlab.com/libeigen/eigen/-/commit/591906477bc8c8102dbefceefe10d81648865394

maxiwell commented:
Hi @cantonios, sorry for the delay, I was on vacation. I changed the `configure.py` to add this flag when `ld` is 2.35 or greater. Would it be better to tag someone here to help us review?

cantonios commented on configure.py:
```diff
@@ -1331,6 +1331,20 @@ def main():
 
+  if is_ppc64le():
+    # Check if the linker has MMA Dynamic Dispatch support
+    ld_version = run_shell(['ld', '--version']).split()
```
Is a raw `ld` necessarily the linker?

mihaimaruseac replied:
I don't think we're using `ld` everywhere.

mihaimaruseac replied:
I think that's much better, though we have builds where gcc is not the default compiler either.

maxiwell replied:
This Dynamic Dispatch feature only works with gcc. So, I updated my patch to check that.
